[FIX] web: show warning before closing view form

nle-odoo opened the pull request:
In a view form, a onchange could only be triggered just before closing
the modal. Hence people could not be made aware of a warning which may
have changed what they were about to do.

This change only allow to save a view form when the warning is closed.

opw-642067

beledouxdenis commented:
It does solves the issue.

Nevertheless, relying on the fact the warning dialog is displayed it not the best solution, in my opinion.

Would that not be possible that an onchange returning a warning set  `save_obj.error` ? Just asking, if this is too much complicated or lead to too much change, don't.

nle-odoo commented:
I updated the pull request, now when the warning is displayed the save_list objects are set to error (the only location where save list is filled is in `save: function(){}`).

beledouxdenis commented:
This approach is better :). However, it works for the popup case (in the sales order lines popup form view) but not for the case of the editable list and saving directly.

nle-odoo commented:
if I add a warning to the `lunch.order.line` onchange:

```diff
diff --git a/addons/lunch/lunch.py b/addons/lunch/lunch.py
index 1d17b44..125d1cb 100644
--- a/addons/lunch/lunch.py
+++ b/addons/lunch/lunch.py
@@ -331,8 +331,12 @@ class lunch_order_line(osv.Model):
     def onchange_price(self, cr, uid, ids, product_id, context=None):
         if product_id:
             price = self.pool.get('lunch.product').browse(cr, uid, product_id, context=context).price
-            return {'value': {'price': price}}
-        return {'value': {'price': 0.0}}
+            res = {'value': {'price': price}}
+        else:
+            res = {'value': {'price': 0.0}}
+        if product_id == 6:
+            res['warning'] = {'title':'yeah', 'message':'bambo'}
+        return res
 
     def order(self, cr, uid, ids, context=None):
         """ 
```

 then click on save, i get the warning and the record is not saved:

![2015-06-18-111821_543x256_scrot](https://cloud.githubusercontent.com/assets/9977887/8227994/b7911f72-15ab-11e5-8d86-aa5352962547.png)

so maybe I'm missing something?

beledouxdenis commented:
- Create a new sale order
 - Create a new line, make sure to have the editable list and not the popup (it depends on user groups).
 - Choose "Ipad mini", 5000 as quantity, save directly.

nle-odoo commented:
okay, in my browser it happend only sometimes and it never happens when I added a console.log at [view_form.js#L583](https://github.com/odoo-dev/odoo/blob/8.0-warning-view-form-nle/addons/web/static/src/js/view_form.js#L583) line.

I guess that it changes something in event order or something, and my fix only worked for : 

#### saving record :
a) start of saving record
b) add record to save_list
c) wait mutex (onchange, ...)
d) do the saving

#### do onchange :

e) start onchange
f) end onchange, display warning
g) close warning

when __f)__ happens between __b)__ and __d)__. but sometimes __f)__ must happen before __b)__.

so I changed the fix so that instead whenever a warning is open, the close operation should be cancelled, so now to block the save : 
 *  __f)__ should be done before __d)__
 *  __g)__ should be done after __d)__

(so the real saving will not be done when the warning popup is displayed) 

which thanks to mutex etc should be ok in all case.

nle-odoo commented:
@beledouxdenis I think it's either this or the first jquery css selector

beledouxdenis commented:
>okay, in my browser it happend only sometimes and it never happens when I added a console.log at view_form.js#L583 line.

Maybe there is a race condition somewhere. And these aren't easy to track down. 

Regarding this fix: To me, it's almost identical then your first suggestion, you simply check if there is warning popup displayed, you are just doing it differently, not with the dom.

In such a case, I rather like your first suggestion, this one being fare more complicated and risky (I am not sure the "closing" event of the popup is called for each possibility regarding popups closing, with escape, automatically, and so on).

Therefore, maybe re-suggest your first prop. I will then ask Olivier if this is acceptable or not.

nle-odoo commented:
ok, I replaced c711f47badf with 1a9c0977d9

the first commit was d024d7d but I think there was an issue (the record would still be saved in some case.

beledouxdenis commented:
Unfortunately, Olivier agrees with me that this is not clean, we shouldn't prevent the save on the fact the warning popup is within the DOM.

We will therefore give a new look to find another cleaner patch.
